### DISCUSSION TITLE
How to identify which version of Yarn is used by a project?

### DISCUSSION THREAD

**Discussion by gmoniava** [yarnpkg/berry#5591] [Q&A]
If I understood correctly [this](https://yarnpkg.com/getting-started/migration), suggests that say if my project was configured using yarn 1, and then I upgrade yarn to version 2, and use it in this project, it can cause issues.
Is that correct?

If yes, imagine a situation where I download a repo, and say I have some version of yarn installed, should not there be a way to tell which version of yarn was that project configured with, to avoid potential problems?

Similar question has been asked [here](https://stackoverflow.com/questions/72346056/how-to-identify-which-version-of-yarn-is-used-by-a-project).

**Comment by rally25rs** [comment]
Someone else might be able to articulate this better, but;

With yarn modern, it is a common practice to commit the yarn binary into the code repo (often at `.yarn/releases/`in your repo), and specify its path using [yarnPath](https://yarnpkg.com/configuration/yarnrc#yarnPath) in the config. That way everyone uses the same version of yarn for that project, and different projects can use different version of yarn without relying on everyone having to keep their local version up to date.

Newer versions of yarn v1 also respect the `yarnPath` setting, so a code repo can be set up with a yarnPath that points to a yarn modern release, and running a `yarn` command would run your globally installed yarn v1, which would then see the `yarnPath` and go load that yarn runtime instead (which could be yarn v3 or whatever).

So to answer your question:

> a way to tell which version of yarn was that project configured with

Check for a `yarnPath` in the config. If a repo is set up that way, then you should be fine. Whatever version of yarn it is set to will be used.

If it is a repo not set up that way, or an older yarn v1 repo, then as far as I know there isn't a way to tell the exact yarn version that was used when creating the project initially.

**Comment by polarathene** [comment]
As already answered `yarnPath` may be common to encounter, but the preferred way going forward is to have `package.json` declare the project `packageManager`, which will also use semver to ensure you're using the correct version.

NodeJS provides CorePack and only requires you to enable it `corepack enable` once on your system. That will create shims so that when the yarn command is run it'll use the version identified in `package.json`, if you don't have it corepack will install it for you. Outside of the project it'll fallback to your global yarn version. It works well.

More information here: https://github.com/yarnpkg/berry/discussions/5744#discussioncomment-7065602
